### DISCUSSION TITLE
Long-term degradation simulation of battery takes too long. Is there any reason why this problem occurs?

### DISCUSSION THREAD

**Discussion by CaritasPostech** [pybamm-team/PyBaMM#2653] [Q&A] [answered]
I would like to use PyBaMM to calculate capacity degradation over long cycles using a DFN model and measure how long it takes.
Since I have long-term capacity fade experiment data of over 4000 cycles, I wrote the following code to test PyBaMM using the same number of cycles.

===================================================
# Profile 1

Profile = [(f"Charge at 0.5C until 4.2V"    
       f"Hold at 4.2V until 145mA"
    "Rest for 10 minutes",
    f"Discharge at 0.5C until 3V"
    "Rest for 10 minutes")
    ]*4000
    
MODEL = pybamm.lithium_ion.DFN({"SEI": "solvent-diffusion limited"})

var_pts = {
    "x_n": 20,  # negative electrode
    "x_s": 20,  # separator 
    "x_p": 20,  # positive electrode
    "r_n": 30,  # negative particle
    "r_p": 30,  # positive particle
}

def EXP_DFN_SEI(PF_num, MODEL, var_pts):
    param = pybamm.ParameterValues("OKane2022")
    pybamm.set_logging_level("NOTICE")
    param.set_initial_stoichiometries(1);
    print("%d"%(PF_num))

    experiment = pybamm.Experiment(Profile[PF_num])
    sim = pybamm.Simulation(MODEL, parameter_values=param, experiment=experiment, var_pts=var_pts)
    sim.solve(calc_esoh=False)

EXP_DFN_SEI(PF_num, MODEL, var_pts)

=================================================================

However, in the case of PyBaMM simulation, one cycle is calculated in a short time (4-5 sec) at the beginning, but as it goes on, it takes a longer time (about 40 sec) to calculate one cycle. Is there a reason why it slows down as the number of cycles increases, or is there a problem with my code?

If there is anyone who knows the reason, I would really appreciate a reply. Thank you for your help.

**Comment by TomTranter** [comment, marked as answer]
Sounds like a bug or memory leak. How many cycles does it take to slow down significantly? In order to debug the problem a bit can you change to a faster model to see if the problem persists and log the console output to a file. Also does stopping after say 100 cycles then restarting the simulation from the previous solution help at all?

**Reply by CaritasPostech** [reply]
In my code currently, the simulation time is gradually increasing as the cycles progress. It seems that the memory leak you mentioned is also currently occurring. Initially, the memory occupancy was less than 20%, but as the simulation progresses, it rises to 97% (I have confirmed that the same problem occurs on a PC with 64GB RAM, as my PC has 32GB RAM). I will try rewriting the code as you mentioned, breaking it down every 100 cycles. Thank you for your kind answer.

**Reply by brosaplanella** [reply]
You can also use the `save_at_cycles` option in the solve step to only save some cycles, which should free some memory.

**Reply by Dapid** [reply]
I don't think it is a memory leak, but that it is intentionally saving a lot of information. As brosaplanella says, `save_at_cycles` is the solution.
